ACCUMULO-626 Iterator fuzzer/tester

joshelser opened the pull request:
Some early code for automatically testing iterators.

madrob commented on iterator-test-harness/pom.xml:
```diff
@@ -0,0 +1,56 @@
+    <dependency>
+      <groupId>org.reflections</groupId>
+      <artifactId>reflections</artifactId>
+    </dependency>
```
Do we need to add this to the NOTICE?

joshelser replied:
uhhh license is http://www.wtfpl.net/. Ideally, this will be ripped out before it's committed.

madrob replied:
LEGAL-135 and http://www.apache.org/legal/resolved.html#category-a indicate that it's fine. I just wanted to know if we need to document the new dependency somehow.

joshelser replied:
No worries, your attention to that detail is appreciated. It was really just a "google, copy/paste" job to keep me moving on that I didn't mean to keep long-term. :)

ctubbsii replied:
LICENSE/NOTICE include things that we distribute, not things which satisfy our dependencies. So, no, I don't think we need to include anything extra there, unless we're including it's source in our project, or expecting to bundle its jar in our -bin.tar.gz

As for the WTFPL, it's fine... I remember looking into this a bit for [ACCUMULO-1496](https://issues.apache.org/jira/browse/ACCUMULO-1496). That license even permits the code to be freely re-licensed, so even if we did include it, we could include it under the Apache License instead anyway.

madrob replied:
> That license even permits the code to be freely re-licensed, so even if we did include it, we could include it under the Apache License instead anyway.

This came up in LEGAL-135 and the community felt it was better to leave as is, instead of re-licensing.

keith-turner commented on iterator-test-harness/src/main/java/org/apache/accumulo/iteratortest/IteratorTestCaseFinder.java:
```diff
@@ -0,0 +1,61 @@
+   */
+  public static List<IteratorTestCase> findAllTestCases() {
+    log.info("Searching {}", IteratorTestCase.class.getPackage().getName());
+    Reflections reflections = new Reflections(IteratorTestCase.class.getPackage().getName());
```
If this is a new dependency, maybe you could use Guava's [ClassPath.getAllClasses()][1].   Could iterator an filter based on type.

[1]: http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/reflect/ClassPath.html#getAllClasses%28%29

joshelser replied:
Yeah, like I hinted at above, this was a quick hack to get this functionality and I hope to remove it (despite being a very nice library to use). Ideally, I'd not need any extra dependencies to do it (currently don't have any deps outside of accumulo-core, junit and slf4j that I recall), and I'd prefer to leave it that way. If it's hard to do with regular reflection libs, Guava sounds like a nice choice. Thanks for the recommendation.

keith-turner commented on iterator-test-harness/src/test/java/org/apache/accumulo/iteratortest/WholeRowIteratorTest.java:
```diff
@@ -0,0 +1,147 @@
+    List<Value> values = new ArrayList<>();
+
+    // Generate the output data from the input data
+    for (Entry<Key,Value> entry : INPUT_DATA.entrySet()) {
```
Using [RowIterator](https://accumulo.apache.org/1.6/apidocs/org/apache/accumulo/core/client/RowIterator.html) may simplify this code a bit.

wjsl commented on iterator-test-harness/src/main/java/org/apache/accumulo/iteratortest/IteratorTestUtil.java:
```diff
@@ -0,0 +1,44 @@
+  }
+
+  public static SortedKeyValueIterator<Key,Value> createSource(IteratorTestInput input) {
+    return new ColumnFamilySkippingIterator(new SortedMapIterator(Objects.requireNonNull(input).getInput()));
```
Might be good to make another iterator here (or wrapping the SMI) that re-uses the Key and Value instances returned by the getters. Storing the output of a source in a list or map without making a copy is a common pitfall.

I had a small iterator benchmark that I modified to printout the identify of each Key and Value, and currently notice the same Value instance has the potential to be re-used. I will try to verify if Keys exhibit the same behavior-- I have heard anecdotes that they do.

keith-turner replied:
b. loss did in this in TransformingIteratorTest.  See line [#662 in 1.6.4](https://github.com/apache/accumulo/blob/1.6.4/core/src/test/java/org/apache/accumulo/core/iterators/user/TransformingIteratorTest.java#L662)

joshelser replied:
> Storing the output of a source in a list or map without making a copy is a common pitfall.

Ooo, that's a good one I didn't think to test. Sounds like a good idea to implicitly test. I can also see a test that checks that Keys and values aren't reused (with a reference check).

joshelser replied:
Turns out, this was a really good suggestion. Helped me find a bug in my test cases (need to make sure to copy the Key-Value being emitted). The Tablet Scanner does this when converting the Key and Value from the iterator stack into a `KeyValue`. Thanks again, Bill!

keith-turner commented on iterator-test-harness/src/main/java/org/apache/accumulo/iteratortest/testcases/ReSeekTestCase.java:
```diff
@@ -0,0 +1,107 @@
+        Key reSeekStartKey = skvi.getTopKey();
+
+        // Deepcopy the iterator
+        skvi = skvi.deepCopy(new SimpleIteratorEnvironment());
```
creating a completely new iterator instance and initializing it would be more realistic

keith-turner commented:
There may be value in running some of these test against an iterator stack and not just a single iterator.

keith-turner commented:
For iterators that can be run at major compaction, how can we exercise delete keys?

joshelser commented:
> For iterators that can be run at major compaction, how can we exercise delete keys?

Very good question! The completely mocked-out IteratorEnvironment is a sign that I haven't thought past the scan case. I'm not even sure if the ColumnFiltering iterator with SortedMapIterator is even a 100% valid in-memory reproduction of how it works in a TabletServer.

I had left input entirely up to the user, so I'm not sure of a simple way to test deletions without potentially messing up some expected table structure. There are some tricky edge cases at compaction that I haven't yet considered how to encapsulate. Combiners on non-full compactions are another good example.

keith-turner commented on iterator-test-harness/src/main/java/org/apache/accumulo/iteratortest/testcases/DeepCopyTestCase.java:
```diff
@@ -0,0 +1,63 @@
+
+      SortedKeyValueIterator<Key,Value> copy = skvi.deepCopy(new SimpleIteratorEnvironment());
+
+      copy.seek(testInput.getRange(), Collections.<ByteSequence> emptySet(), false);
```
Seeking one deep copy should have no impact on other deep copies.   Would be nice to verify this.  Could interleave reading and seeking a few deep copies.

joshelser commented:
Rebased on top of master and addressed the current comments:

* Removed Reflections dependency
* Reuse Key/Value in the test iterator stack
* New deepCopy test that iterates over multiple copies.
* More accurate re-seek test case.

Only outstanding thing I want to do before an initial commit is to run all of the iterators in {{o.a.a.c.iterators}} through this framework in accumulo-test (start the dog-fooding).

Keith had some suggestions about fancier testing (iterator stacks, compaction/scopes, and delete keys), which I'm not sure I want to try to solve right away. I feel like I've let these changes sit for far too long already. I'd be happy to wait if anyone else is interested in working on it short-term, though.

joshelser commented:
Drat, another thing I just thought of: this is meant to be user facing, and I think it should eventually hit public API. It might be premature to put it there immediately (since this hasn't been vetted by anyone but myself), but does anyone have thoughts on a path to "graduate" into public API status?

dhutchis commented on iterator-test-harness/src/main/java/org/apache/accumulo/iteratortest/testcases/DeepCopyTestCase.java:
```diff
@@ -0,0 +1,63 @@
+    try {
+      skvi.init(source, testInput.getIteratorOptions(), new SimpleIteratorEnvironment());
+
+      SortedKeyValueIterator<Key,Value> copy = skvi.deepCopy(new SimpleIteratorEnvironment());
```
Some thoughts on another test-- call deepCopy in the middle of an iteration, in addition to the beginning after init() but before any next() calls.  Some iterators may do this in order to "save" the iteration at a certain point.

joshelser replied:
Thanks for taking a look, Dylan!

> call deepCopy in the middle of an iteration, in addition to the beginning after init() but before any next() calls

Yeah, there are lots of good tests we can do around deepCopy. I'll see if I can add one for this specific case. Please feel free to send a PR if you definitely want to see this in the initial changeset :)

>  Some iterators may do this in order to "save" the iteration at a certain point.

But that's not something the "system" does, is it? I can see the value, but I'm not sure if that's something that would happen through "normal" Accumulo use.

dhutchis replied:
Ah I see Josh-- is the purpose to test only system iterators, as opposed to generic tests to check user-defined iterators?  That makes more sense.  

Some users may define iterators that expect a column family, whereas all the tests seek with `Collections.<ByteSequence> emptySet(), false`.  There may be many other places where a user does something "non-normal" intentionally.  If we make this public API, we should state that user iterators do not necessarily need to pass all the IteratorTestCases, or at least that failing an IteratorTestCase does not imply a bug.

dhutchis replied:
Further along that note, many user iterators leave deepCopy() as `throw new UnsupportedOperationException()`.  They would fail many of these IteratorTestCases.

joshelser replied:
> is the purpose to test only system iterators, as opposed to generic tests to check user-defined iterators? That makes more sense. 

Yep. First and foremost, my goal was to provide a framework for users to test iterators without running them through Accumulo (since it's often hard to know when you're hitting some of the edge cases). However, I can totally see it growing into a framework to catch "common pitfalls" like you outlined :)

> Some users may define iterators that expect a column family, whereas all the tests seek with Collections.<ByteSequence> emptySet(), false

Yeah, this is something I'm not quite sure how to handle colfams (I'm throwing that into the same bucket as the compaction stuff and more that Keith pointed out).

>  If we make this public API, we should state that user iterators do not necessarily need to pass all the IteratorTestCases, or at least that failing an IteratorTestCase does not imply a bug.

Very good point. We don't really have an "evolving" annotation. Maybe we just talk about it and the lack of public API declares it as evolving intrinsically?

keith-turner commented:
> but does anyone have thoughts on a path to "graduate" into public API status?

Well currently iterators are not declared as public API.

joshelser commented:
> Well currently iterators are not declared as public API.

Ooo, that is a good point (which I often forget). Perhaps I'll just table the idea for now and we can address if/when we do something with iterators.

joshelser commented:
Okie doke. Pulled in Dylan's test improvements (thanks again for that!). I think the only thing I have left that I want to do is to wire up as many of the existing iterators we provide for users with some tests in the accumulo-test module.

Any concerns, speak up soon. I'll likely merge over the weekend unless I hear objection to not do so. Then, I will file some follow-on JIRA work for the lingering improvements.
